dialdbg/src/main.rs

```diff
@@ -0,0 +1,239 @@
+    uri: Option<String>,
+}
+
+fn read_config(config: PathBuf) -> Result<(String, String)> {
+    let (uri, secret): (String, String);
+    let r: Value = from_str(fs::read_to_string(config)?.as_str())?;
+    if let Some(u) = r.get("address") {
+        // We use some funky as_str -> to_string logic as serde_json weirdly wraps to_string'ed
+        // Values in quotes. Also, we want to verify "address" and "secret" are actually strings.
+        // Was I too lazy to use regular serde? Yes.
+        if let Some(uri_str) = u.as_str() {
+            uri = uri_str.to_string();
+        } else {
+            return Err(anyhow!(
+                "expected 'address' value to be a string in provided config"
+            ));
+        }
+    } else {
+        return Err(anyhow!("no top-level 'address' value in provided config"));
+    }
+    if let Some(s) = r.get("secret") {
+        if let Some(sec_str) = s.as_str() {
+            secret = sec_str.to_string();
+        } else {
+            return Err(anyhow!(
+                "expected 'secret' value to be a string in provided config"
+            ));
+        }
+    } else {
+        return Err(anyhow!("no top-level 'secret' value in provided config"));
+    }
+    Ok((uri, secret))
+}
+
+async fn dial_grpc(uri: &str, secret: &str) {
+    let creds = dial::RPCCredentials::new(
+        None,
+        "robot-location-secret".to_string(),
+        secret.to_string(),
+    );
+
+    // `connect` may propagate an error here; log the error with a prefix so we can still
+    // process logs and not immediately return from the main function.
+    if let Err(e) = dial::DialOptions::builder()
+        .uri(uri)
+        .with_credentials(creds)
+        .disable_webrtc()
+        .allow_downgrade()
+        .connect()
+        .await
+    {
+        log::error!("{}: {e}", strings::DIAL_ERROR_PREFIX);
+    }
+}
+
+async fn dial_webrtc(uri: &str, secret: &str) -> Option<stats::StatsReport> {
+    let creds = dial::RPCCredentials::new(
+        None,
+        "robot-location-secret".to_string(),
+        secret.to_string(),
+    );
+
+    // `connect` may propagate an error here; log the error with a prefix so we can still
+    // process logs and not immediately return from the main function. Assuming there was
+    // no error, return the stats report of the underlying RTCPeerConnection.
+    match dial::DialOptions::builder()
+        .uri(uri)
+        .with_credentials(creds)
+        .allow_downgrade()
+        .connect()
+        .await
+    {
+        Ok(c) => match c {
+            ViamChannel::WebRTC(c) => Some(c.get_stats().await),
+            _ => None,
+        },
+        Err(e) => {
+            log::error!("{}: {e}", strings::DIAL_ERROR_PREFIX);
+            None
+        }
+    }
+}
+
+fn output_connection_stats(stats: stats::StatsReport, out: &mut Box<dyn io::Write>) -> Result<()> {
+    // StatsReport contains 13 types of stat reports; there may be more relevant stats to print
```

Comment:
would implement Display trait here instead so we can make it easier to use elswhere

Reply:
I could do this, but to clarify I would need to make a new type that wraps `stats::StatsReport` and implement `Display` for that type (both the struct and trait are foreign).

Reply:
if you do it like `pub struct StatsReport(stats::StatsReport)`